From f83e3209d2e16419d4e4a27f9d3ed694f4c2f738 Mon Sep 17 00:00:00 2001
From: nyanmisaka <nst799610810@gmail.com>
Date: Wed, 27 Dec 2023 22:54:43 +0800
Subject: [PATCH 04/37] HACK: lavu: allow image copy for NV15 and NV20

Signed-off-by: nyanmisaka <nst799610810@gmail.com>
---
 libavutil/imgutils.c | 26 +++++++++++++++++++-------
 1 file changed, 19 insertions(+), 7 deletions(-)

diff --git a/libavutil/imgutils.c b/libavutil/imgutils.c
index 9ab5757cf6..7a18247c31 100644
--- a/libavutil/imgutils.c
+++ b/libavutil/imgutils.c
@@ -341,14 +341,12 @@ int av_image_check_sar(unsigned int w, unsigned int h, AVRational sar)
     return AVERROR(EINVAL);
 }
 
-static void image_copy_plane(uint8_t       *dst, ptrdiff_t dst_linesize,
-                             const uint8_t *src, ptrdiff_t src_linesize,
-                             ptrdiff_t bytewidth, int height)
+static void image_copy_plane2(uint8_t       *dst, ptrdiff_t dst_linesize,
+                              const uint8_t *src, ptrdiff_t src_linesize,
+                              ptrdiff_t bytewidth, int height)
 {
     if (!dst || !src)
         return;
-    av_assert0(FFABS(src_linesize) >= bytewidth);
-    av_assert0(FFABS(dst_linesize) >= bytewidth);
     for (;height > 0; height--) {
         memcpy(dst, src, bytewidth);
         dst += dst_linesize;
@@ -356,6 +354,19 @@ static void image_copy_plane(uint8_t       *dst, ptrdiff_t dst_linesize,
     }
 }
 
+static void image_copy_plane(uint8_t       *dst, ptrdiff_t dst_linesize,
+                             const uint8_t *src, ptrdiff_t src_linesize,
+                             ptrdiff_t bytewidth, int height)
+{
+    if (!dst || !src)
+        return;
+    av_assert0(FFABS(src_linesize) >= bytewidth);
+    av_assert0(FFABS(dst_linesize) >= bytewidth);
+    image_copy_plane2(dst, dst_linesize,
+                      src, src_linesize,
+                      bytewidth, height);
+}
+
 void av_image_copy_plane_uc_from(uint8_t *dst, ptrdiff_t dst_linesize,
                                  const uint8_t *src, ptrdiff_t src_linesize,
                                  ptrdiff_t bytewidth, int height)
@@ -424,7 +435,8 @@ void av_image_copy(uint8_t *dst_data[4], int dst_linesizes[4],
                    enum AVPixelFormat pix_fmt, int width, int height)
 {
     ptrdiff_t dst_linesizes1[4], src_linesizes1[4];
-    int i;
+    int i, is_nv15_20 = pix_fmt == AV_PIX_FMT_NV15 ||
+                        pix_fmt == AV_PIX_FMT_NV20;
 
     for (i = 0; i < 4; i++) {
         dst_linesizes1[i] = dst_linesizes[i];
@@ -432,7 +444,7 @@ void av_image_copy(uint8_t *dst_data[4], int dst_linesizes[4],
     }
 
     image_copy(dst_data, dst_linesizes1, src_data, src_linesizes1, pix_fmt,
-               width, height, image_copy_plane);
+               width, height, is_nv15_20 ? image_copy_plane2 : image_copy_plane);
 }
 
 void av_image_copy_uc_from(uint8_t *dst_data[4], const ptrdiff_t dst_linesizes[4],
-- 
2.25.1

